### PR TITLE
Add events on http routing

### PR DESCRIPTION
**Pull request by Aschen**
## What does this PR do ?

This PR add 6 events related to the HTTP method used. `http:get`, `http:post`, `http:put`, `http:patch`, `http:delete`, `http:head` (`http:options` was already present).  
See [Jira#KZL-42](https://jira.kaliop.net/browse/KZL-42) or https://github.com/kuzzleio/kuzzle/issues/1099 for more context.  

The events are fired in `Router.route()` and now this method always return a promise.  

Changes are documented here : https://github.com/kuzzleio/documentation/pull/478

### How should this be manually tested?

  - Step 1 : Run a Kuzzle stack with plugins debug : `DEBUG=kuzzle:plugins docker-compose -f docker-compose/dev.yml up`
  - Step 2 : Try running a curl request and verify if the correct event is fired : `curl -X PUT localhost:7512`
image

  - Step 3 : Repeat with all HTTP methods: `methods=(GET POST PUT PATCH DELETE HEAD); for method in $methods; do curl -X $method localhost:7512/_getInfo; done`

### Boyscout

 - Fix listener arguments on `unhandledRejection` (Related to #1020 )
 - Fix `should().be.called()` syntax in tests
 - Allow user to remove `node_modules` created by docker without `sudo`
 - Replace references to back-office by admin console

### CODE REVIEW

**Comment by Aschen** [comment]
I noticed that there were no functional test for the plugin system.  
Is it on purpose or just due to lack of time ?

File: lib/api/core/httpRouter/index.js [hunk 1]
```diff
@@ -158,48 +158,56 @@ class Router {
           .catch(error => replyWithError(cb, request, error));
       }
 
-      return replyWithError(cb, request, new BadRequestError(`Unrecognized HTTP method ${httpRequest.method}`));
+      return Promise.reject(replyWithError(cb, request, new BadRequestError(`Unrecognized HTTP method ${httpRequest.method}`)));
```
Comment (ballinette) [inline lib/api/core/httpRouter/index.js]: what is the reason to reject the promise ?

File: lib/api/core/httpRouter/index.js [hunk 2]
```diff
@@ -158,48 +158,56 @@ class Router {
...
-        return replyWithError(cb, request, new NotFoundError(`API URL not found: ${routeHandler.url}`));
-      }
+      return this.kuzzle.pluginsManager.trigger(httpEvent, request)
+        .then(mutatedRequest => {
+          routeHandler.request = mutatedRequest;
 
-      if (httpRequest.content.length <= 0) {
-        return routeHandler.invokeHandler(cb);
-      }
+          if (routeHandler.handler === null) {
```
Comment (ballinette) [inline lib/api/core/httpRouter/index.js]: should be moved before the event trigger, to avoid calling plugin pipes if the URL does not match any route.

**Comment by ballinette** [comment]
@Aschen what do you mean by "functional test for the plugin system" ?

File: lib/api/core/httpRouter/index.js (hunk 1 above)
Comment (Aschen) [reply lib/api/core/httpRouter/index.js]: To ensure consistency on return value. This function always return a promise, either fulfilled or rejected but maybe it's not mandatory ?

File: lib/api/core/httpRouter/index.js (hunk 2 above)
Comment (Aschen) [reply lib/api/core/httpRouter/index.js]: Are you sure ? Even if the request does not match any route, it's still a HTTP request received by Kuzzle, shouldn't we trigger the event anyway ?

**Comment by Aschen** [comment]
@ballinette I mean tests with cucumber to ensure the events can be hooked or piped to a plugin

File: docker-compose/scripts/install-plugins.sh [hunk 3]
```diff
@@ -12,7 +12,7 @@ for target in ${plugins_dir}/* ${protocols_dir}/* ; do
     echo 'Installing dependencies for ' $(basename "$target")
     cd "$target"
     npm install --unsafe
+    chmod 777 -R node_modules
```
Comment (alexandrebouthinon) [inline docker-compose/scripts/install-plugins.sh]: Why 777 ?

File: docker-compose/scripts/run-dev.sh [hunk 4]
```diff
@@ -6,6 +6,7 @@ elastic_host=${kuzzle_services__db__client__host:-http://elasticsearch:9200}
 
 npm install --unsafe
 npm rebuild all --unsafe
+chmod 777 -R node_modules
```
Comment (alexandrebouthinon) [inline docker-compose/scripts/run-dev.sh]: Why 777 ?

File: docker-compose/scripts/run-test.sh [hunk 5]
```diff
@@ -6,6 +6,7 @@ elastic_host=${kuzzle_services__db__client__host:-http://elasticsearch:9200}
 
 npm install --unsafe
 npm install --unsafe --only=dev
+chmod 777 -R node_modules
```
Comment (alexandrebouthinon) [inline docker-compose/scripts/run-test.sh]: Why 777 ?

**Comment by alexandrebouthinon** [comment]
Use chown instead of chmod for permissions :wink:

File: docker-compose/scripts/install-plugins.sh (hunk 3 above)
Comment (Aschen) [reply docker-compose/scripts/install-plugins.sh]: Because I'm lazy.. I will fix that :)

**Comment by Aschen** [comment]
Seen with @alexandrebouthinon , we can't use `chown` because we are inside the docker container and we don't know the host user uid. We can assume it's 1000 but since we are not sure I prefer to use a permissive `chmod`

File: lib/api/core/httpRouter/index.js [hunk 6]
```diff
@@ -158,48 +158,58 @@ class Router {
...
-
-      if (httpRequest.content.length <= 0) {
-        return routeHandler.invokeHandler(cb);
-      }
-
-      if (httpRequest.headers['content-type']
-        && !httpRequest.headers['content-type'].startsWith('application/json')) {
-        return replyWithError(cb, request, new BadRequestError(`Invalid request content-type. Expected "application/json", got: "${httpRequest.headers['content-type']}"`));
-      }
+      return this.kuzzle.pluginsManager.trigger(httpEvent, request)
```
Comment (scottinet) [inline lib/api/core/httpRouter/index.js]: These new events must be triggered **after** the request has been validated. It makes no sense sending a request to plugins if an URL is invalid or if a request is malformed.

File: lib/api/core/httpRouter/index.js (hunk 1 above)
Comment (ballinette) [reply lib/api/core/httpRouter/index.js]: No, it doesn't. This method works with callbacks and not promises (see also comment from scottinet)

**Comment by Aschen** [comment]
Very clever solution @ballinette, I like it :)

**Comment by benoitvidis** [comment]
@ballinette I do not agree with your solution to an almost non-existing problem.
Injecting some host-specific data just for this looks quite ugly to me. 

In any case, this is for dev only and not blocking imo.

**Comment by Aschen** [comment]
I revert the change suggested by @ballinette because travis wasn't happy

**Comment by ballinette** [comment]
hmmm not blocking, I agree, except that I'm not fan of having a script that silently change the filemod on my host... :/

but travis should work if we fix the command in  kuzzle/test/travis-bin/run-tests.sh like this:
```
DEV_UID=`id -u` DEV_GID=`id -g` docker-compose up -f docker-compose/test.yml run kuzzle 
```

**Pull request merged**